diff --git a/Eigen/src/Core/util/DisableStupidWarnings.h b/Eigen/src/Core/util/DisableStupidWarnings.h
index fe0cfec0b..d317d236e 100755
--- a/Eigen/src/Core/util/DisableStupidWarnings.h
+++ b/Eigen/src/Core/util/DisableStupidWarnings.h
@@ -74,25 +74,25 @@
 #endif
 
 #if defined __NVCC__
-  #pragma diag_suppress boolean_controlling_expr_is_constant
-  // Disable the "statement is unreachable" message
-  #pragma diag_suppress code_is_unreachable
-  // Disable the "dynamic initialization in unreachable code" message
-  #pragma diag_suppress initialization_not_reachable
-  // Disable the "invalid error number" message that we get with older versions of nvcc
-  #pragma diag_suppress 1222
-  // Disable the "calling a __host__ function from a __host__ __device__ function is not allowed" messages (yes, there are many of them and they seem to change with every version of the compiler)
-  #pragma diag_suppress 2527
-  #pragma diag_suppress 2529
-  #pragma diag_suppress 2651
-  #pragma diag_suppress 2653
-  #pragma diag_suppress 2668
-  #pragma diag_suppress 2669
-  #pragma diag_suppress 2670
-  #pragma diag_suppress 2671
-  #pragma diag_suppress 2735
-  #pragma diag_suppress 2737
-  #pragma diag_suppress 2739
+  // #pragma diag_suppress boolean_controlling_expr_is_constant
+  // // Disable the "statement is unreachable" message
+  // #pragma diag_suppress code_is_unreachable
+  // // Disable the "dynamic initialization in unreachable code" message
+  // #pragma diag_suppress initialization_not_reachable
+  // // Disable the "invalid error number" message that we get with older versions of nvcc
+  // #pragma diag_suppress 1222
+  // // Disable the "calling a __host__ function from a __host__ __device__ function is not allowed" messages (yes, there are many of them and they seem to change with every version of the compiler)
+  // #pragma diag_suppress 2527
+  // #pragma diag_suppress 2529
+  // #pragma diag_suppress 2651
+  // #pragma diag_suppress 2653
+  // #pragma diag_suppress 2668
+  // #pragma diag_suppress 2669
+  // #pragma diag_suppress 2670
+  // #pragma diag_suppress 2671
+  // #pragma diag_suppress 2735
+  // #pragma diag_suppress 2737
+  // #pragma diag_suppress 2739
 #endif
 
 #else
