diff --git a/faiss/gpu/impl/CuvsIVFFlat.cu b/faiss/gpu/impl/CuvsIVFFlat.cu
index 2cccee860..c4cb13f48 100644
--- a/faiss/gpu/impl/CuvsIVFFlat.cu
+++ b/faiss/gpu/impl/CuvsIVFFlat.cu
@@ -427,13 +427,11 @@ void CuvsIVFFlat::copyInvertedListsFrom(const InvertedLists* ivf) {
     // Precompute the centers vector norms for L2Expanded distance
     if (this->metric_ == faiss::METRIC_L2) {
         cuvs_index->allocate_center_norms(raft_handle);
-        raft::linalg::rowNorm(
+        raft::linalg::rowNorm<raft::linalg::L2Norm, true>(
                 cuvs_index->center_norms().value().data_handle(),
                 cuvs_index->centers().data_handle(),
                 cuvs_index->dim(),
                 (uint32_t)nlist,
-                raft::linalg::L2Norm,
-                true,
                 raft_handle.get_stream());
     }
 }
